### PR TITLE
Refactor BrailleHandler.setDisplayByName

### PR DESCRIPTION
**Pull request by LeonarddeR**
### Link to issue number:
None

### Summary of the issue:
braille.BrailleHandler.setDisplayname is too complex. It also relies on reentrancy and is difficult to grasp.

### Description of user facing changes
See bug fixes below

### Description of development approach
* setDisplayByName is now split in three functions
* setDisplayByName will no longer set noBraille as a fallback if noBraille is already active. This avoids reinitializing noBraille multiple times
* BrailleHandler._detectionEnabled is removed, as it was supposed to be equal to the boolean value of BrailleHandler._detector and that value was checked anyway. Note that that's supposed to be private API
* When configuration was set to auto detect, BrailleHandler.handleDisplayUnavailable relied on behavior in setDisplayByName to re-enable the detector when falling back to no braille. This logic is now simplified.

### Testing strategy:
- [x] Ensure that auto detection works over USB
- [x] Ensure that auto detection works over Bluetooth
- [x] Ensure that when a display was detected over Bluetooth, the detector is still running for USB in order to switch when an USB connection is active
- [x] Ensure that auto detection will be activated when a display looses connectivity and autodetect is in config

### Known issues with pull request:
None known

### Change log entries:
Bug fixes
- NVDA will no longer unnecessarily switch to no braille multiple times during auto detection, mainly resulting in a cleaner log and less overhead. (#14524)
- NVDA will now switch back to USB if a HID Bluetooth device (such as the HumanWare Brailliant or APH Mantis) is automatically detected and an USB connection becomes available. This only worked for Bluetooth Serial ports before. (#14524)

### Code Review Checklist:

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] Security precautions taken.

### CODE REVIEW

File: source/braille.py [hunk 1]
```diff
@@ -348,6 +349,9 @@
 #: braille displays should be automatically detected and used.
 #: @type: str
 AUTO_DISPLAY_NAME = AUTOMATIC_PORT[0]
+#: The name of the noBraille display driver
+#: @type: str
+NO_BRAILLE_DISPLAY_NAME = "noBraille"
```
Comment (seanbudd) [inline source/braille.py:354]: I think new lines make this more readable.
The type hint is unnecessary

```suggestion

#: The name of the noBraille display driver
NO_BRAILLE_DISPLAY_NAME: str = "noBraille"

```

File: source/braille.py [hunk 2]
```diff
@@ -18,6 +18,7 @@
 	Set,
 	Tuple,
 	Union,
+	Type
```
Comment (seanbudd) [inline source/braille.py:21]: add comma

```suggestion
	Type,
```

File: source/braille.py [hunk 3]
```diff
@@ -1830,7 +1834,11 @@ def getFocusRegions(
 	from NVDAObjects import NVDAObject
 	if isinstance(obj, CursorManager):
 		region2 = (ReviewTextInfoRegion if review else CursorManagerRegion)(obj)
-	elif isinstance(obj, DocumentTreeInterceptor) or (isinstance(obj,NVDAObject) and NVDAObjectHasUsefulText(obj)): 
+	elif (
+		isinstance(obj, DocumentTreeInterceptor) or (
+			isinstance(obj, NVDAObject) and NVDAObjectHasUsefulText(obj)
+		)
+	):
```
Comment (seanbudd) [inline source/braille.py:1841]: More readable indentation style

```suggestion
	elif (
		isinstance(obj, DocumentTreeInterceptor)
		or (
			isinstance(obj, NVDAObject)
			and NVDAObjectHasUsefulText(obj)
		)
	):
```

File: source/braille.py [hunk 4]
```diff
@@ -1959,92 +1966,99 @@ def _get_enabled(self):
...
-	# Note: when working on setDisplayByName, look for opportunities to simplify
-	# and move logic out into smaller helper functions.
-	def setDisplayByName(  # noqa: C901
+	def setDisplayByName(
 			self,
 			name: str,
-			isFallback=False,
+			isFallback: bool = False,
 			detected: typing.Optional[bdDetect.DeviceMatch] = None,
 	):
```
Comment (seanbudd) [inline source/braille.py:1974]: can you add a return type here? (bool)

File: source/braille.py [hunk 5]
```diff
@@ -1959,92 +1966,99 @@ def _get_enabled(self):
...
+				self._setDisplay(fallbackDisplayClass, isFallback=False)
+			return False
 
+	def _switchDisplay(
+			self,
+			oldDisplay: Optional["BrailleDisplayDriver"],
+			newDisplayClass: Type["BrailleDisplayDriver"],
+			**kwargs
+	) -> "BrailleDisplayDriver":
+		sameDisplayReinit = newDisplayClass == oldDisplay.__class__
```
Comment (seanbudd) [inline source/braille.py:2017]: `ReInit` seems more readable than `Reinit`

```suggestion
		sameDisplayReInit = newDisplayClass == oldDisplay.__class__
```

File: source/braille.py [hunk 6]
```diff
@@ -2379,31 +2403,30 @@ def handleDisplayUnavailable(self):
...
-		self.setDisplayByName("noBraille", isFallback=True)
+		newDisplay = (
+			AUTO_DISPLAY_NAME
+			if config.conf["braille"]["display"] == AUTO_DISPLAY_NAME
+			else NO_BRAILLE_DISPLAY_NAME
+		)
+		self.setDisplayByName(newDisplay, isFallback=True)
 
-	def _enableDetection(self, usb=True, bluetooth=True, keepCurrentDisplay=False, limitToDevices=None):
+	def _enableDetection(self, usb=True, bluetooth=True, limitToDevices=None):
```
Comment (seanbudd) [inline source/braille.py:2413]: can this get type hints?

**Review by seanbudd — COMMENTED** [review]
Hi @leonardder, the approach looks good to me, just minor suggestions.

Due to the risk of this PR, I think we should merge it after #14512

File: source/braille.py (hunk 4 above)
Comment (leonardder) [reply source/braille.py:1974]:
```suggestion
	) -> bool:
```

**Comment by leonardder** [comment]
@seanbudd Thanks for the review.
I removed the change to handlePostConfigProfileSwitch as I realized that every display supporting port selection will have its port defined in config, also when it is automatic. Given no complaint about this anyway, let's leave it as is.

**Review by seanbudd — APPROVED** [review]
Thanks, LGTM, will merge once we branch for release

File: source/braille.py [hunk 7]
```diff
@@ -2078,98 +2088,99 @@ def _handleEnabledDecisionFalse(self):
...
-			self._disableDetection()
+		except Exception:
+			# For auto display detection, logging an error for every failure is too obnoxious.
+			if not detected:
+				log.error(f"Error initializing display driver {name!r}", exc_info=True)
+			elif bdDetect._isDebug():
+				log.debugWarning(f"Couldn't initialize display driver {name!r}", exc_info=True)
+			fallbackDisplayClass = _getDisplayDriver(NO_BRAILLE_DISPLAY_NAME)
+			# Only initialize the fallback if it is not already set
+			if self.display.__class__ == fallbackDisplayClass:
```
Comment (michaelDCurran) [inline source/braille.py:2129]: Should this not be `!=` instead?
Getting reports if NVDA can not initialize a specifically configured display on start-up, then handler.display is None and the Braille page in the GUI is broken.

**Pull request merged**